internal/handler/control.go

```diff
@@ -1,22 +1,151 @@
 // -*- Mode: Go; indent-tabs-mode: t -*-
 //
 // Copyright (C) 2017-2018 Canonical Ltd
+// Copyright (C) 2020 IOTech Ltd
 //
 // SPDX-License-Identifier: Apache-2.0
 
 package handler
 
 import (
+	"context"
 	"fmt"
+	"io"
+	"net/http"
+	"regexp"
+	"sync"
+	"time"
 
+	"github.com/edgexfoundry/device-sdk-go/internal/cache"
 	"github.com/edgexfoundry/device-sdk-go/internal/common"
+	"github.com/edgexfoundry/device-sdk-go/pkg/models"
+	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+	"github.com/google/uuid"
 )
 
-func DiscoveryHandler(requestMap map[string]string) {
-	common.LoggingClient.Info(fmt.Sprintf("service: discovery request"))
+type discoveryLocker struct {
+	busy bool
+	id   string
+	mux  sync.Mutex
 }
 
+var locker discoveryLocker
+
 func TransformHandler(requestMap map[string]string) (map[string]string, common.AppError) {
 	common.LoggingClient.Info(fmt.Sprintf("service: transform request: transformData: %s", requestMap["transformData"]))
 	return requestMap, nil
 }
+
+func DiscoveryHandler(w http.ResponseWriter) {
+	locker.mux.Lock()
+	if locker.id == "" {
+		locker.id = uuid.New().String()
+	}
+	locker.mux.Unlock()
+
+	if w != nil {
+		msg := fmt.Sprintf("Discovery triggered or already running, id = %s", locker.id)
+		w.WriteHeader(http.StatusAccepted) //status=202
+		_, _ = io.WriteString(w, msg)
+	}
+
+	// locker is used to prevent duplicate call of protocol specific device discovery
+	// locker would be released once the DS main thread received the candidate devices
+	// found by discovery
+	locker.mux.Lock()
+	defer locker.mux.Unlock()
+	if locker.busy {
+		common.LoggingClient.Info(fmt.Sprintf("Device discovery process is running, id = %s", locker.id))
+		return
+	}
+	locker.busy = true
+	common.LoggingClient.Info(fmt.Sprintf("service %s discovery triggered", common.ServiceName))
+
+	ctx := context.WithValue(context.Background(), common.CorrelationHeader, locker.id)
+	deviceCh := make(chan []models.DiscoveredDevice)
+	go common.Discovery.Discover(deviceCh)
+	go filterAndAdd(ctx, deviceCh)
```

Comment:
If the service gets shut down, where will the resources for these goroutines get cleaned up?

Reply:
now that we're using go-mod-bootstrap, we can leverage the ctx to stop the goroutine and release the resource.